### ISSUE TITLE
we need a better concept than staged/unstaged

### ISSUE THREAD

**Issue by bpoldrack** [datalad/datalad#1289]
As discussed in last hangout, in direct mode there are no unstaged changes.
In general the fact, that we cannot distinguish staged and unstaged changes in direct mode should lead to not use this idea with annex repos at all.

But while trying to rework the Repo classes' `dirty()` methods and the `ok_clean_git` test helper in PR #1277 this is raising questions. There is at least one important place, we use this: `interface/utils.py` in `save_dataset`. There we want to commit if we have either explicitly given paths to commit or there is something staged. As of https://github.com/datalad/datalad/pull/1277/commits/d59cbd5eb4b3f678e654623899a78b6ae901fc29 I keep the current behaviour in indirect mode, therefore still allowing for the parameter `working_tree` for `dirty()` (triggers a warning instead of an exception).
By now this means kind of a correct behaviour, since `dirty(index=True, working_tree=False)` returns `True` whenever there is something staged. It's just different what is staged in direct mode and that's okay - it's direct mode, it should behave like direct mode.

BUT: While the technical behaviour might be okay-ish, it's troublesome to provide it that way. We encourage the idea of distinguishing, what for an annex isn't a valid distinction. As long as we have it that way, the user of an `AnnexRepo` has to be aware of what that means in different annex modes. I don't see us (or even future contributors) having this is mind all the time.

So, I think we need a concept of how to catagorize possible changes in an annex repository in a more fitting way. The minimum would be to rename the parameters (originally coming from GitPython) from `index, working_tree` to `staged, unstaged`, which is still troublesome but at least better than talking about a working tree, that isn't there.
What else can you think of? Perhaps consider especially the situation described in `save_dataset`. Any other conceptional approach on what kind of changes to commit than "staged" ones? Please note, that it has not necessarily to fit into `dirty()`. Once we have an idea, what we are looking for, we can have a dedicated method to do so, of course.

**Comment by mih** [comment]
How exactly is `save_dataset` a problem? IIRC it explicitly does not care about "untracked" files. But staged and explicitly given path only.

**Comment by bpoldrack** [comment]
It does not care for untracked files, yes. But it also tries to not care for unstaged changes. In order to do so, it is calling `dirty(index=True, working_tree=False, untracked_files=False)`, which is possible only, if we generally allow to use `working_tree` (or similar parameter).

**Comment by bpoldrack** [comment]
At this point of `save_dataset` we kind of want to stick to the concept of staged changes, while on the other hand we know, that this very concept isn't actually valid for annexes. I'm looking for a consistent way to approach this.

**Comment by mih** [comment]
No, it sets that parameter because it wants to avoid the test. Remove it, and let it always avoid the test, and it will remain happy.

Even with working_tree=True, it would not commit anything untracked, so the problem here has no consequences.

**Comment by mih** [comment]
A different approach would be to not test at all in `save_dataset`. The only reason it is doing the test in the present code is to avoid the spectacular fireworks should anyone dare to commit a change where there is none ;-)

**Comment by bpoldrack** [comment]
Well, if this is the case, it never worked exactly as intended, I guess. ;-)
Because `working_true=False` does not just avoid the test, but gives a different result, if there are unstaged changes (which would be committed, if we change it to `with working_tree=True`.

But okay, if we actually care for untracked files only, we can just remove it, of course.

**Comment by bpoldrack** [comment]
Hah - would NOT be committed - commit() would fail instead, of course. Okay, got it. :-D

**Comment by mih** [comment]
Worth reiterating: Often the present approach of "outside" testing is the result of the behavior of the low-level layers to whine as hard and long as possible (e.g. #1111, ...). Many of these tests must be redundant, as either git-annex or git will do (some of) that again. Moving towards a result-based error behavior would avoid many issues like this and has the potential to speed things up, and make fewer bugs from fewer lines of code and fewer misinterpreted test conditions. I.e. If I tell the machine to save `fileX` and there is nothing to be saved, because there is no change, it is not an error, because the result is what I wanted -- all saved.

I am not saying that it should not communicate back that it did nothing, but there is a difference between a litany of log ERRORs and `return False` or `return None`. Which would naturally go away, if we assumed that it is OK to call commit, even if there is a chance that there is nothing to commit.

**Comment by bpoldrack** [comment]
Yes, agree.

Okay, so while `save_dataset` probably can just be called without `working_tree` (testing the approach right now), I'll keep this open, until the corresponding PR is done. Just in case I find more ...

**Comment by mih** [comment]
I adjusted title and top-comment to document that there is no concrete evidence for a staged/unstaged problem yet.
